### ISSUE TITLE
Human client crashes when opening production screen

### ISSUE THREAD

**Issue by Vezzra** [freeorion/freeorion#1109]
### Environment

* **FreeOrion Version**: 93847d84df206d6a28b2ba1919fec522ec7feba6
* **Operating System**: OSX
* **Graphic card used**: AMD Radeon HD 6490M 256 MB
* **Fetched as**
  * Compiled from source

### Description

On OSX, when opening the production screen the human client throws an exception, see these lines of `freeorion.log`:
```
2016-11-15 22:55:57.632665 [debug] Client : ProductionWnd.cpp:999 : ProductionWnd::UpdateQueue()
2016-11-15 22:55:57.679054 [debug] Client : ProductionWnd.cpp:999 : ProductionWnd::UpdateQueue()
2016-11-15 22:55:57.730758 [debug] Client : ClientNetworking.cpp:317 : ClientNetworking::NetworkingThread() : Networking thread will be terminated due to disconnect exception "Operation canceled"
2016-11-15 22:55:58.731709 [debug] Client : Process.cpp:231 : Process::Impl::Kill calling kill(m_process_id, SIGKILL)
2016-11-15 22:55:58.731924 [debug] Client : Process.cpp:233 : Process::Impl::Kill calling waitpid(m_process_id, &status, 0)
2016-11-15 22:55:58.738236 [debug] Client : Process.cpp:235 : Process::Impl::Kill done
2016-11-15 22:55:58.738307 [debug] Client : HumanClientFSM.cpp:664 : (HumanClientFSM) ~PlayingTurn
2016-11-15 22:55:58.738337 [debug] Client : HumanClientFSM.cpp:334 : (HumanClientFSM) ~PlayingGame
2016-11-15 22:55:58.859604 [error] Client : chmain.cpp:304 : main() caught exception(std::exception): Layout::DoLayout() : calculated column positions do not sum to the width of the layout
```
Curiously, this doesn't happen *always*. Apparently, depending on what I do before I go to the production screen, the crash happens or not.

I haven't been able to reproduce the crash on Windows.

### Expected Result

Well, no crash... 😉 

### Steps to reproduce

The following procedure has so far reliably reproduced the crash:

* Start new game.
* Open research screen.
* Open production screen.
* Crash.

**Comment by MatGB** [comment]
This might be related and I got called away before I could open issue but I'm finding planets aren't selecting properly to start building immediately when I open prod or change system within it, have to select the planet and build lists are sometimes for a previous system. 

will write up properly tomorrow when I'm back on desktop if it's not.

**Comment by geoffthemedio** [comment]
I couldn't reproduce as described on Windows, but I quickstarted a game, then set my two scouts to explore, pressed turn ~5 times and got a crash. Was able to reproduce immediately after, with debugger attached. Stack:

```
>   GG::Wnd::SizeMove(const GG::Pt & ul_=<struct at NULL>, const GG::Pt & lr_=<struct at NULL>) Line 531    C++
    GG::DeferredLayout::PreRender() Line 54 C++
    `anonymous namespace'::SystemRow::Init() Line 739   C++
    `anonymous namespace'::SystemRow::PreRender() Line 744  C++
    GG::EventPumpBase::LoopBody(GG::GUI * gui=0x00083ade, GG::EventPumpState & state={...}, bool do_non_rendering=255, bool do_rendering=true) Line 80  C++
    GG::ModalEventPump::operator()() Line 112   C++
    GG::SDLGUI::Run() Line 890  C++
    mainSetupAndRun() Line 286  C++
```

Or, putting a breakpoint on the throw line in Layout::DoLayout:

```
>   GG::Layout::DoLayout(GG::Pt ul={...}, GG::Pt lr={...}) Line 431 C++
    GG::DeferredLayout::PreRender() Line 54 C++
    `anonymous namespace'::SystemRow::Init() Line 739   C++
    `anonymous namespace'::SystemRow::PreRender() Line 744  C++
    GG::EventPumpBase::LoopBody(GG::GUI * gui=0x00083b29, GG::EventPumpState & state={...}, bool do_non_rendering=255, bool do_rendering=true) Line 80  C++
    GG::ModalEventPump::operator()() Line 112   C++
    GG::SDLGUI::Run() Line 890  C++
```

with locals:

```
+       this    0x42780007 {m_cells={ size=??? } m_border_margin=??? m_cell_margin=??? ...} GG::Layout *
+       ul  {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       lr  {x={m_value=1115160585 } y={m_value=29 } }  GG::Pt
        greatest_usable_min_over_stretch_ratio  0.00000000000000000 double
+       row_effective_min_usable_sizes  { size=1 }  std::vector<unsigned int,std::allocator<unsigned int> >
        greatest_min_over_stretch_ratio 0.00000000000000000 double
+       column_effective_min_usable_sizes   { size=2 }  std::vector<unsigned int,std::allocator<unsigned int> >
+       new_min_size    {x={m_value=1115160585 } y={m_value=5 } }   GG::Pt
+       original_size   {x={m_value=352 } y={m_value=22 } } GG::Pt
        min_due_to_margin   2   unsigned int
        min_due_to_margin   2   unsigned int
+       min_space_needed    {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       margin  {x={m_value=1 } y={m_value=0 } }    GG::Pt
+       min_usable_size {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       assignment  {m_old_val=false m_t=??? }  GG::ScopedAssign<bool>
+       assignment  {m_old_val=false m_t=??? }  GG::ScopedAssign<bool>
+       new_parent_min_size {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       ul  {x={m_value=2 } y={m_value=-859946519 } }   GG::Pt
+       ul_margin   {x={m_value=127 } y={m_value=29 } } GG::Pt
+       lr_margin   {x={m_value=127 } y={m_value=29 } } GG::Pt
+       lr  {x={m_value=1 } y={m_value=0 } }    GG::Pt
+       window_size {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       resize_ul   {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       available_space {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       min_size    {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       resize_lr   {x={m_value=0 } y={m_value=0 } }    GG::Pt
+       min_usable_size {x={m_value=1115160585 } y={m_value=5 } }   GG::Pt
```

which suggests DeferredLayout::m_lr_prerender hasn't been initialized properly.

**Comment by LGM-Doyle** [comment]
@MatGB I can see the selection problem.  I will look into it.  I don't know if it is related to the crashes.

@Vezzra, I have not seen these crashes.  Could you capture a stack trace so that I can look into the problem?

**Comment by Vezzra** [comment]
I just spent hours trying to narrow down the commit which introduced this bug and failed. The bug is extremely elusive. It happens frequently enough to make playing a pain, but can't be reliably reproduced, so sometimes I have to try several times before the crash happens again.

Which renders `git bisect` practically useless :frowning_face: 

Finally tried to reproduce the bug with current master (06e6e82b34ad202cb54c6ae6a04d5f0b64dd9e75) and haven't been able to, no matter what I did.

This is quite frustrating. I have to call it a day for now, TBC on Sunday, hopefully will be able to provide a backtrace then.

**Comment by LGM-Doyle** [comment]
@Vezzra don't worry about the backtrace.  If you can play with the debugger attached and wait for the crash, then that may be sufficient.

However, I had not refreshed my browser and I missed @geoffthemedio's post until right now.  I can't replicate it but there is information there that I can use to find a solution. It is frustrating to not be able to replicate the crash myself.

**Comment by Vezzra** [comment]
I can't get a backtrace, even with the debugger attached. :frowning_face: Maybe I'm doing something wrong, here is what I do:
- Launch the debugger (`lldb` on OSX)
- Launch FO
- Start a default quickstart game
- Attach to the human client process via issueing `attach FreeOrion` on the debugger prompt
- This pauses FO, issueing `c` resumes execution
- Play until FO crashes
- This produces the following output: `Process 2155 exited with status = 0 (0x00000000)`
- Issueing `bt` gives me `error: invalid thread`

Well, crap :frowning_face:

Apparently the process exits cleanly after catching that exception, and after that `lldb` doesn't give me anything about the no longer running process. What can I do to get that backtrace???

I can't pin down the commit which introduced the issue (because of it's erratic behaviour), I can't get a backtrace, and you can't reproduce it on your system.

On top of that, this is becoming really serious over here. I've experienced these crashes several times now when trying to test some of the pending PRs, not only when I try to open the production screen, but also simply when doing a few turns (like @geoffthemedio described).

This seriously impairs my ability to do any dev and testing work. As well as making it more or less pointless to provide test builds (playing is practically impossible).

I'm quite screwed at the moment...

**Comment by LGM-Doyle** [comment]
Now you can see why I'm opposed to catching, but not handling exceptions.  Technically, freeorion has not crashed, but it has also discarded the backtrace information that might help solve the problem.

I have two suggestions: first modify your executables to preserve the stack and second make sure that you attach to all of the executables in the unlikely case it is not the human client that is crashing. 

First, with luck modifying the executables will be easy because the linux version are already configured to keep the stacktrace.  The three executables freeorion (human client), freeorionca (AI client) and freeoriond (server) are in freeorion/client/human/chmain.cpp, freeorion/client/AI/camain.cpp and freeorion/server/dmain.cpp respectively.  

You will see at the top of each file at section like:

```
    #if defined(FREEORION_LINUX)
    /* Freeorion aims to have exceptions handled and operation continue normally.
    An example of good exception handling is the exceptions caught around config.xml loading.
    After catching and informing the user it continues normally with the default values.

    An exception that can not be handled should allow freeorion to crash and keep
    a complete stack trace of the intial exception.
    Some platforms do not support this behavior.

    When FREEORION_CHMAIN_KEEP_BACKTRACE is defined, do not catch an unhandled exceptions,
    unroll and hide the stack trace, print a message and still crash anyways. */
    #define FREEORION_CHMAIN_KEEP_STACKTRACE
    #endif
```

Change `FREEORION_LINUX` to `FREEORION_MACOSX` in all three files and if all goes well you will have a version the crashes and leaves a backtrace.

Second, try to either attach to all executable processes at once, or each type in turn.  lldb may only be catching the backtrace from the process that it is attached to.  lldb may only be able to attach to one process at a time, I don't know.  

From the [cheatsheet](http://lldb.llvm.org/lldb-gdb.html) that I linked to, it looks like the command to attach to a process is `process attach --pid <number>`.  Use `ps =elf | grep freeorion` or `pgrep --list-name freeorion` to figure out what the pids are each time you run freeorion. You can make sure that you have attached correctly by intentionally creating a crash in the process that you think you have attached to.

Good luck.

If my instructions were not clear, ask questions and I will try a different explanation.

I am working on fixing the production window/SidePanel selection problem, which might be related to this problem and could be done today.

**Comment by geoffthemedio** [comment]
Can you not configure the debugger to pause on exceptions being thrown? MSVC can...

**Comment by Vezzra** [comment]
> If my instructions were not clear, ask questions and I will try a different explanation.

@LGM-Doyle, thanks for the pointers, your instructions have been perfectly clear 😃

Enabling the `*_KEEP_STACKTRACE` defines worked smoothly. I've been able to attach to the human client, server and one AI process, but only the human client yielded a backtrace (and was indeed the process that crashed, the others just got shut down in the cleanup process).

Here is the backtrace:
```
* thread #1: tid = 0x7937, 0x9804669a libsystem_kernel.dylib`__pthread_kill + 10, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x9804669a libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x9b098634 libsystem_pthread.dylib`pthread_kill + 101
    frame #2: 0x91fe4de6 libsystem_c.dylib`abort + 156
    frame #3: 0x933202f9 libc++abi.dylib`abort_message + 169
    frame #4: 0x93343483 libc++abi.dylib`default_terminate_handler() + 272
    frame #5: 0x93340ac0 libc++abi.dylib`std::__terminate(void (*)()) + 14
    frame #6: 0x933404db libc++abi.dylib`__cxa_throw + 122
    frame #7: 0x0057f00f FreeOrion`GG::Layout::DoLayout(GG::Pt, GG::Pt) + 7001
    frame #8: 0x005d2837 FreeOrion`GG::DeferredLayout::PreRender() + 93
    frame #9: 0x0056e2f3 FreeOrion`GG::GUI::PreRenderWindow(GG::Wnd*) + 47
    frame #10: 0x00589509 FreeOrion`GG::ListBox::PreRender() + 321
    frame #11: 0x0053625f FreeOrion`GG::DropDownList::PreRender() + 237
    frame #12: 0x0056e2f3 FreeOrion`GG::GUI::PreRenderWindow(GG::Wnd*) + 47
    frame #13: 0x0056eafa FreeOrion`GG::GUI::PreRender() + 50
    frame #14: 0x0053ee42 FreeOrion`GG::EventPumpBase::LoopBody(GG::GUI*, GG::EventPumpState&, bool, bool) + 614
    frame #15: 0x0053efc5 FreeOrion`GG::ModalEventPump::operator()() + 81
    frame #16: 0x005c0912 FreeOrion`GG::SDLGUI::Run() + 46
    frame #17: 0x00058d0c FreeOrion`mainSetupAndRun() + 821
    frame #18: 0x0036ad45 FreeOrion`-[GUIController applicationDidFinishLaunching:] + 23
    frame #19: 0x9ac8d0f3 Foundation`__57-[NSNotificationCenter addObserver:selector:name:object:]_block_invoke + 50
    frame #20: 0x93eefc34 CoreFoundation`__CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 20
    frame #21: 0x93dcf901 CoreFoundation`_CFXNotificationPost + 3713
    frame #22: 0x9ac6e2c4 Foundation`-[NSNotificationCenter postNotificationName:object:userInfo:] + 92
    frame #23: 0x988c2868 AppKit`-[NSApplication _postDidFinishNotification] + 365
    frame #24: 0x988c2558 AppKit`-[NSApplication _sendFinishLaunchingNotification] + 239
    frame #25: 0x988bf05d AppKit`-[NSApplication(NSAppleEventHandling) _handleAEOpenEvent:] + 827
    frame #26: 0x988be92a AppKit`-[NSApplication(NSAppleEventHandling) _handleCoreEvent:withReplyEvent:] + 277
    frame #27: 0x9b0d5119 libobjc.A.dylib`-[NSObject performSelector:withObject:withObject:] + 84
    frame #28: 0x9aca4e57 Foundation`__76-[NSAppleEventManager setEventHandler:andSelector:forEventClass:andEventID:]_block_invoke + 118
    frame #29: 0x9aca499f Foundation`-[NSAppleEventManager dispatchRawAppleEvent:withRawReply:handlerRefCon:] + 451
    frame #30: 0x9aca4794 Foundation`_NSAppleEventManagerGenericHandler + 211
    frame #31: 0x9bbea94e AE`aeDispatchAppleEvent(AEDesc const*, AEDesc*, unsigned long, unsigned char*) + 583
    frame #32: 0x9bbb5467 AE`dispatchEventAndSendReply(AEDesc const*, AEDesc*) + 44
    frame #33: 0x9bbb535a AE`aeProcessAppleEvent + 313
    frame #34: 0x94bdf1c1 HIToolbox`AEProcessAppleEvent + 55
    frame #35: 0x988b82ee AppKit`_DPSNextEvent + 2707
    frame #36: 0x988b71e5 AppKit`-[NSApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 350
    frame #37: 0x988abb9c AppKit`-[NSApplication run] + 907
    frame #38: 0x98820fa0 AppKit`NSApplicationMain + 2082
    frame #39: 0x0005b315 FreeOrion`main + 599
    frame #40: 0x00022ad5 FreeOrion`start + 53
```
However, and this is important, I have *not* been able to reproduce the crash with current master (4f31dad024f6dce7278520ef16c1839984aed00e). The commit I built to reproduce the crash to get the backtrace is b14aa84bfd3851a8530f9f7fb07a0a7c39d5192d.

Can't tell you if something that has been committed in the meantime really fixed the issue or just hid it. Hopefully the backtrace helps a bit getting to the bottom of this.

@geoffthemedio, can you reproduce the crash you got with current master?

**Comment by Vezzra** [comment]
> Can you not configure the debugger to pause on exceptions being thrown? MSVC can...

Maybe, don't know, I'm not familiar with the `lldb` debugger. Just looked up the few commands I needed to attach to running processes and getting the backtrace.

Unfortunately I can't use the Xcode IDE for debugging purposes (because trying to launch FO from within Xcode crashes it), so I've to use the debugger on the command line.

**Comment by geoffthemedio** [comment]
@Vezzra I cannot

**Comment by LGM-Doyle** [comment]
I'm glad it worked.

It is also possible, if OSX has a crash reporter that with `*_KEEP_STACKTRACE` enabled the crash reporter would have generated a backtrace on its own.

I do not think that the crash has fixed itself.  However, both of the stacks provided indicate a problem in the `DropDownList` code.

The `DropDownList` is built around a standard `ListBox` which is uses as the drop down list in a separate `ModalEventPump` from the main loop, while moving the selected row from the list to the anchor area of the `DropDownList` to render the selected row twice.  I added `PreRender()` and then fixed that size accounting for limiting the drop down.  I'm searching through that code for a mistake.

**Comment by LGM-Doyle** [comment]
@geoff I've never used `lldb`, so I just picked the first command from the cheatsheet that I linked to that could handle this problem.  `gdb` can do as you described, so I imagine `lldb` can do it as well.

**Comment by Vezzra** [comment]
I can't test if #1124 actually does fix the underlying issue as the problem I reported here hasn't been reproducible for several commits now. As #1124 apparently fixes #1125, I go ahead assuming it has fixed the issue reported here too, so will close as resolved.

If the issue turns up again, I'll reopen or create a new issue.

**Comment by geoffthemedio** [comment]
You could revert to a commit that is known to have the issue, then cherry pick the changes in the PR to test.

**Comment by Vezzra** [comment]
Well, yes, but even that can only reliably tell me if the fix does *not* work (as in, if the issue persists, then the fix obviously does not work). But for that I can test the PR as it is.

The problem is, apparently the issue can disappear just because something in the code changed. So if a build containing the fix proposed here works (which it does), I'm still not able to tell if that's because the PR actually *did* fix the underlying issue, or if it's just by chance.

So I decided to just go with the assumption that the issue has been fixed correctly, wait and see if it turns up again, and reopen in case that happens.

**Comment by geoffthemedio** [comment]
Knowing that the fix didn't work would probably be useful, though?

**Comment by Vezzra** [comment]
I admit being reluctant because of too much hassle. I've already racked up a considerable amount of frustration trying to track down this very annoying crash. Each check/test run is very tedious. I have to clean and completely rebuild every time (because doing otherwise changes the results), which of course takes a lot of time.

Anyway, I've tried reverting to a commit that is known to have the issue and cherry picking the commits of the PR too, which gave me a conflict with the particular commit I had chosen (probably because I went back too far), meaning I'd now have to go hunting for a commit that has the issue but no conflicts with any of the commits if the PR... ugh.

I know this is being a bit lazy, but I decided to just be content with hoping the fix actually did address the underlying issue, and just wait if it resurfaces...
